lib/SILGen/SILGenExpr.cpp

```diff
@@ -2293,6 +2293,20 @@ RValue RValueEmitter::visitRebindSelfInConstructorExpr(
   return SGF.emitEmptyTupleRValue(E, C);
 }
 
+static bool isNullableTypeInC(Type ty) {
+  if (ty->hasReferenceSemantics())
+    return true;
+  
+  auto &C = ty->getASTContext();
+  auto nom = ty->getAnyNominal();
+  if (!nom)
+    return false;
+
+  return nom == C.getUnsafePointerDecl()
+    || nom == C.getUnsafeMutablePointerDecl()
+    || nom == C.getAutoreleasingUnsafeMutablePointerDecl();
+}
```

Comment:
@jrose-apple I updated the `mayLieAboutOptionalReturn` code to check for pointer types and factored out the `isNullableTypeInC` check into a separate helper function. This look good?